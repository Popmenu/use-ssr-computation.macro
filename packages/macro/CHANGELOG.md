# Change Log

## [1.2.4] - 2023-10-02
* Support Subscription computations
* Drop support for async computations

## [1.2.0] - 2023-08-28
* Support async computations
* Add support for error handling

## [1.1.0] - 2023-08-06
* Add support for `skip` option that skips the computation

## [1.0.3] - 2023-07-21
* Split the package into `macro` and `runtime` packages

[1.2.4]: https://github.com/shakacode/use-ssr-computation.macro/compare/%40popmenu/use-ssr-computation.macro%401.2.3...%40popmenu/use-ssr-computation.macro%401.2.4
[1.2.0]: https://github.com/shakacode/use-ssr-computation.macro/pull/51/files
[1.1.0]: https://github.com/shakacode/use-ssr-computation.macro/pull/40
[1.0.3]: https://github.com/shakacode/use-ssr-computation.macro/tree/f4edc2717a19a0a332db5cc679c0d536253d19ab
