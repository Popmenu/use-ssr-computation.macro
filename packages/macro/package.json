{
  "name": "@popmenu/use-ssr-computation.macro",
  "version": "1.1.0",
  "main": "lib/index.macro.js",
  "types": "lib/index.macro.d.ts",
  "license": "MIT",
  "scripts": {
    "build": "tsc -p .",
    "build:watch": "yarpm build --watch",
    "test": "jest",
    "compile-example": "yarn clear:babel-cache && babel --plugins=babel-plugin-macros --presets=@babel/preset-typescript example/index.ts",
    "clear:babel-cache": "rimraf -rf ./node_modules/.cache/babel-loader/*",
    "clean": "rimraf lib",
    "prepublishOnly": "yarpm clean && yarpm build"
  },
  "files": [
    "lib"
  ],
  "homepage": "https://github.com/Popmenu/use-ssr-computation.macro",
  "bugs": {
    "url": "https://github.com/Popmenu/use-ssr-computation.macro/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/Popmenu/use-ssr-computation.macro.git"
  },
  "author": "Roman Kuksin <rkuksin.cpp@gmail.com>",
  "dependencies": {
    "@popmenu/use-ssr-computation.runtime": "*",
    "babel-plugin-macros": "^3.0.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.8.4",
    "@babel/preset-typescript": "^7.22.5",
    "@types/babel-plugin-macros": "^3.1.0",
    "@types/babel__core": "^7.1.3",
    "@types/jest": "^29.5.2",
    "@types/node": "^20.3.1",
    "babel-plugin-tester": "^11.0.4",
    "jest": "^29.3.1",
    "rimraf": "^5.0.1",
    "ts-jest": "^29.1.0",
    "tslint": "^6.1.3",
    "typescript": "^5.1.3",
    "yarpm": "^1.2.0"
  },
  "keywords": [
    "SSR",
    "performance",
    "babel",
    "macro",
    "javascript",
    "typescript"
  ],
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    }
  }
}
