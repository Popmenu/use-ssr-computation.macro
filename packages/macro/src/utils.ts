import * as t from "@babel/types";
import { Options as RuntimeOptions } from "@popmenu/use-ssr-computation.runtime/src/utils"

type MacroOptions = {
  serverSideOnly?: boolean;
  webpackChunkName?: string;
};

export type Options = MacroOptions & RuntimeOptions;

// `ExtractObjectTypes` maps each property of the input object `T` to a string representing its type
type ExtractObjectTypes<T> = {
  [K in keyof Required<T>]: T[K] extends (boolean | undefined) ? 'boolean' :
                    T[K] extends (string | undefined) ? 'string' :
                    never;
};

const macroOptionsToTypes: ExtractObjectTypes<MacroOptions> = {
  serverSideOnly: 'boolean',
  webpackChunkName: 'string',
};

export function extractMacroOptions(optionsNode: t.ObjectExpression): MacroOptions {
  const macroOptions: MacroOptions = {};
  Object.entries(macroOptionsToTypes).forEach(([macroOptionName, macroOptionType]) => {
    for (const property of optionsNode.properties) {
      if (!t.isObjectProperty(property) || !(t.isIdentifier(property.key) || t.isStringLiteral(property.key))) continue;

      const propertyKey = t.isStringLiteral(property.key) ? property.key.value : property.key.name;
      if (propertyKey !== macroOptionName) continue;

      switch (macroOptionType) {
        case 'boolean':
          if (!t.isBooleanLiteral(property.value)) {
            throw new Error(`The ${macroOptionName} property must be a boolean literal.`);
          }
          break;
        case 'string':
          if (!t.isStringLiteral(property.value)) {
            throw new Error(`The ${macroOptionName} property must be a string literal.`);
          }
          break;
      }

      // add to macro options and remove from optionsNode
      macroOptions[macroOptionName] = property.value.value;
      optionsNode.properties.splice(optionsNode.properties.indexOf(property), 1);
    }
  });

  return macroOptions;
};
