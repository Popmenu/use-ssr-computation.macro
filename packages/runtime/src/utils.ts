export type Dependency = number | string | { uniqueId: string; }

export function isDependency(element: any): element is Dependency {
  return typeof element === 'number' || typeof element === 'string' || (typeof element === 'object' && element.uniqueId !== undefined);
}

export const calculateCacheKey = (modulePath: string, dependencies: Dependency[]): string => {
  const dependenciesString = dependencies.map((dependency) => {
    if (typeof dependency === 'number') {
      return dependency.toString();
    }
    if (typeof dependency === 'string') {
      return dependency;
    }
    return dependency.uniqueId;
  }).join(',');

  return `${modulePath}::${dependenciesString}`;
}
