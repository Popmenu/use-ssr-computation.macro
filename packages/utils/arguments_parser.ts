import * as t from "@babel/types";

export type Dependency = number | string | { uniqueId: string; }

export type Options = {
  webpackChunkName?: string;
}

function isDependency(element: any): element is Dependency {
  return typeof element === 'number' || typeof element === 'string' || (typeof element === 'object' && element.uniqueId !== undefined);
}

export function parseSSRComputationArgs(parent: t.CallExpression): { filename: string, dependencies: Dependency[] } {
  const filenameNode = parent.arguments[0];
  const dependenciesNode = parent.arguments[1];

  // Check and parse filename
  if (!t.isStringLiteral(filenameNode)) {
    throw new Error("The first argument must be a path to an existing ts file.");
  }
  const filename: string = filenameNode.value;

  // Check and parse dependencies
  let dependencies: Dependency[] = [];
  if (dependenciesNode) {
    if (!t.isArrayExpression(dependenciesNode)) {
      throw new Error("The second argument must be an array of dependencies.");
    }

    dependencies = dependenciesNode.elements.map((element, index) => {
      if (!isDependency(element)) {
        throw new Error(`Invalid dependency at index ${index}.`);
      }
      return element.value;
    });
  }

  return { filename, dependencies };
}
