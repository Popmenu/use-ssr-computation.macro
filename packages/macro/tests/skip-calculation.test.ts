import { pluginTest } from './utils';

pluginTest('skip calculation when a specific condition is met', `
  import { useSSRComputation } from "../lib/index.macro"
  const skip = true;

  // The option is used to skip the calculation when a specific condition is met
  // The option is necessary because React hooks can't be called conditionally
  // The skip option can take any expression that's get evaluated at runtime
  const x = useSSRComputation("./a.ssr-computation", [], { skip, webpackChunkName: "custom-chunk-name" });
`);
