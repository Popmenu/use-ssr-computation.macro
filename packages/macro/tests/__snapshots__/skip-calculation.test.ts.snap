// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros 1. client-
  The option is used to skip the calculation when a specific condition is met
  The option is necessary because React hooks can't be called conditionally
  The skip option can take any expression that's get evaluated at runtime
: 1. client-
  The option is used to skip the calculation when a specific condition is met
  The option is necessary because React hooks can't be called conditionally
  The skip option can take any expression that's get evaluated at runtime
 1`] = `


import { useSSRComputation } from "../lib/index.macro"
const skip = true;

const x = useSSRComputation("./a.ssr-computation", [], { skip, webpackChunkName: "custom-chunk-name" });


      ↓ ↓ ↓ ↓ ↓ ↓

function _dynamicImport_() {
  return import(
    /* webpackChunkName: "custom-chunk-name-ssr-computations" */ "./a.ssr-computation"
  );
}
import useSSRComputation_Client from "@popmenu/use-ssr-computation.runtime/lib/useSSRComputation_Client";
const skip = true;
const x = useSSRComputation_Client(
  _dynamicImport_,
  [],
  {
    skip,
  },
  "tests/a.ssr-computation",
);

`;

exports[`macros 1. server-
  The option is used to skip the calculation when a specific condition is met
  The option is necessary because React hooks can't be called conditionally
  The skip option can take any expression that's get evaluated at runtime
: 1. server-
  The option is used to skip the calculation when a specific condition is met
  The option is necessary because React hooks can't be called conditionally
  The skip option can take any expression that's get evaluated at runtime
 1`] = `


import { useSSRComputation } from "../lib/index.macro"
const skip = true;

const x = useSSRComputation("./a.ssr-computation", [], { skip, webpackChunkName: "custom-chunk-name" });


      ↓ ↓ ↓ ↓ ↓ ↓

import __a from "./a.ssr-computation";
import useSSRComputation_Server from "@popmenu/use-ssr-computation.runtime/lib/useSSRComputation_Server";
const skip = true;
const x = useSSRComputation_Server(
  __a,
  [],
  {
    skip,
  },
  "tests/a.ssr-computation",
);

`;
